Settings and Teams feature

mr-loop-1 opened the pull request:
closes #210 

- [x] implement all required routes, models and logic following the figma designs
- [x] bug fixes
- [x] get values from config
- [x] rename fn, fields, routes appropriate
- [x] handle error messages and constants
- [x] handle response fields
- [ ] add tests

erenfn commented on backend/migrations/20240916052017-create_teams_table.js:
```diff
@@ -0,0 +1,36 @@
+"use strict";
+
```
This is not the correct way to create the table. Create the table like the others in the models folder, and add it to the index.js file, as done with the others.

mr-loop-1 replied:
ok I will do that, I haven started yet on this.

erenfn commented on backend/src/controllers/user.controller.js:
```diff
@@ -54,5 +54,28 @@ const getCurrentUser = async (req, res) => {
   }
 };
 
+const updateUserDetails = async (req, res) => {
```
Do not include database logic (such as User.findByPk) in the controller file. This file is meant only for checking incoming data. All database operations, like User.findByPk, should be done in the services file. Refer to other flows, such as the files for banners, for guidance.

Take a look at MVCS architecture

mr-loop-1 replied:
Ok I will create a user service. I was looking at other controllers of this file which had queries, will move them to service as well.

erenfn commented on backend/.env:
```diff
@@ -1,9 +0,0 @@
-# Node environment
```
Also this can stay here for now. It allows our tests to run. We'll remove this soon

mr-loop-1 replied:
ok, I will add it again at end of all work right before merge. this will allow me to not have to modify it before every commit.

mr-loop-1 replied:
btw there is another [.env](https://github.com/bluewave-labs/bluewave-onboarding/blob/develop/.env) at root folder, the one I removed is in backend folder

erenfn commented:
> @erenfn I am starting further work on this PR
> 
> this line is in both dev and master branch,
> 
> https://github.com/bluewave-labs/bluewave-onboarding/blob/d97a7a8229d54a17d01d5c812fd3f4441f6d7e87/backend/index.js#L34-L36
> 
> this is making my db clear out after every server start

Try `alter: true` but there are also some problems with it. It is up to you

erenfn commented on backend/src/models/UserTeams.js:
```diff
@@ -0,0 +1,44 @@
+module.exports = (sequelize, DataTypes) => {
```
You don't need this table, it is redundant. You can keep role and  teamId in user table.

mr-loop-1 replied:
I made these more modular so that features like changing teams, as well as choosing invites and multi-team feature can be added later on. I will mark this pr for review soon.

erenfn replied:
@uparkalau Can you take a look at this?

erenfn commented on backend/src/models/Team.js:
```diff
@@ -0,0 +1,40 @@
+module.exports = (sequelize, DataTypes) => {
```
This table is also redundant right now. The only useful feature here is the Team Name, and we can store it instead of teamId in other tables.

mr-loop-1 commented on backend/index.js:
```diff
@@ -31,17 +32,18 @@ sequelize
-  .sync({force:true})
-  .then(() => console.log("Models synced with the database..."))
-  .catch((err) => console.log("Error syncing models: " + err));
+// sequelize
```
will restore this

erenfn replied:
please uncomment this

mr-loop-1 commented:
@erenfn @uparkalau 
The reason I separated team and user is because the application in itself is in context of team (or organisation) and the biggest effect is that multiple users belong to same team.
So, multiple users are logging in to the same context, and hence if I change, say, team name for one user then it has to be for every other one. As per above comment, in absence of teams table, it would mean updating all users.

mr-loop-1 commented:
@erenfn @uparkalau @gorkem-bwl 

I have implemented all the services required according to current designs, as well as routes and controllers needed, although commented some as they are not needed right now and can remove them too.

I added a checklist above in description and It would be helpful if we can be on the same page regarding the overall logic flow before I spend time on the next points. I am happy to answer any comments/doubts/reasoning behind the code.

mr-loop-1 commented:
A few additional points to consider
- whether to soft delete or hard delete a user account
- who can send invite - admin only or anyone in the team
- if there are multiple invites to an email, then after signup from that email, whether first or last invite to be considered

mr-loop-1 commented on backend/migrations/20240906060143-add-name-surname-to-users.js:
```diff
@@ -18,7 +18,6 @@ module.exports = {
     await queryInterface.removeColumn("users", "surname");
     await queryInterface.addColumn("users", "username", {
       type: Sequelize.STRING(50),
-      allowNull: false,
```
it is causing problems in undo

gorkem-bwl commented:
> A few additional points to consider
> 
> * whether to soft delete or hard delete a user account
> * who can send invite - admin only or anyone in the team
> * if there are multiple invites to an email, then after signup from that email, whether first or last invite to be considered

Thanks for those comments. Here are the answers.

1. When a user account is removed, it is hard deleted. All the user information should be purged from the database, including user's name, surname, hashes, and anything related. There should be no PII left in the DB.

2. Only an admin can send an invite. 

3. If there are multiple invites to an email, the last invite can invalidate the previous one, and only the last invite should be valid. In the case of an invalid invite, we can mention this in the login page as "This invitation is invalid", which also would cover wrong invitation codes.

gorkem-bwl commented:
> @erenfn @uparkalau @gorkem-bwl
> 
> I have implemented all the services required according to current designs, as well as routes and controllers needed, although commented some as they are not needed right now and can remove them too.
> 
> I added a checklist above in description and It would be helpful if we can be on the same page regarding the overall logic flow before I spend time on the next points. I am happy to answer any comments/doubts/reasoning behind the code.

WRT the checklist above - can you provide inputs @erenfn / @VladimirParkalov ?

mr-loop-1 commented:
Thanks @gorkem-bwl ,
For point no. 3, is there an invitation code planned on sign up page which is not in designs. Currently, I am having the new user join the last team invite they got on email, or create a default new team for them as admin if there is no invite.

gorkem-bwl commented:
> Thanks @gorkem-bwl ,
> 
> For point no. 3, is there an invitation code planned on sign up page which is not in designs. Currently, I am having the new user join the last team invite they got on email, or create a default new team for them as admin if there is no invite.
> 
> btw the checklist was to track what's done on this pr till now and what's left.

The signup page is only visible when the deployment is done. 

Let me explain. The basic rule of thumb is that, this application is not designed to have a multi-tenant system, but rather deployed on a server. When it is deployed, the first sign up page is shown which is supposed to be filled in by the admin (name, surname, email and password), and then this page will be invisible. 

This ensures that the first signed up person is the admin, and the admin can invite others to sign up. It is not possible for anyone to signup without an invitation. 

Regarding your last sentence: There should be only 1 team, and under no circumstances, the system would create another team. 

Let me know if there is anything unclear.

mr-loop-1 commented:
@gorkem-bwl , so it is one app or organisation per deployment.

gorkem-bwl commented:
> @gorkem-bwl , so it is one app or organisation per deployment (of server and db). I will update the code soon

Sorry for the late answer. This app is supposed to be deployed on a server, which is supposed to run only for one organization.

mr-loop-1 commented:
@gorkem-bwl thanks for clearing the logical flow.

mr-loop-1 commented:
@gorkem-bwl 
I was able to get some time today and changed the logic. I have a few points
- can an admin remove themselves from team
- what happens to team and admin role when the current admin deletes their account
- can there me more than one admin at same time. If yes, can this admin remove other admins

- I noticed that there is a down-arrow selector on role column of team members in figma, does that mean that roles can be changed?

@erenfn @uparkalau
I also wanted to know how should the initialisation of admin happen. My take is a seeder with values of username and password from config or otherwise default.

gorkem-bwl commented:
Thanks for those questions. My answers are inline.

> @gorkem-bwl I was able to get some time today and changed the logic. I have a few points
> 
> * can an admin remove themselves from team

No, admin shouldn't be able to remove themselves from the team. In the future, as long as another admin is present, any  admin should be able to remove themselves.

> * what happens to team and admin role when the current admin deletes their account

The current admin shouldn't be able to delete their account. 

> * can there be more than one admin at same time. If yes, can this admin remove other admins

Yes, there can be more than one admin. An admin can add other admins. An admin can remove other admins.

> * I noticed that there is a down-arrow selector on role column of team members in figma, does that mean that roles can be changed?

Yes, it is a way to change a role. When the drop down selector is clicked, the admin is able to select another auth level for that user. Note that the change shouldn't be instantaneous - it requires the 'Save' button to be clicked afterwards. When the save button is clicked, we better show a notification that this action succeeded (Successfully changed the authorization level). 

Afterwards, the user whose auth level changed should be logging out and logging in to see the changes.

mr-loop-1 commented:
Hi @gorkem-bwl @erenfn @uparkalau 
the pr is ready for review. This now fulfils most of the discussion we've had and the designs too.

The thing not implemented yet is error messages in response when some error occurs where I would need your help after the main logic is approved.

gorkem-bwl commented:
> Hi @gorkem-bwl @erenfn @uparkalau the pr is ready for review. This now fulfils most of the discussion we've had and the designs too.
> 
> The thing not implemented yet is error messages in response when some error occurs where I would need your help after the main logic is approved.

Do you have some sort of UI you can show? (eg video?)

mr-loop-1 commented:
@gorkem-bwl The frontend is not there yet for these changes, I can give a list of features/changes.

gorkem-bwl commented:
> @gorkem-bwl The frontend is not there yet for these changes, I can give a list of features/changes.

Ah ok - nevermind then ;)

erenfn commented on backend/src/controllers/invite.controller.js:
```diff
@@ -0,0 +1,17 @@
+    await inviteService.sendInvite(userId, invitedEmail, role);
+    return res.status(200).json({ message: "Invite Sent Successfully" });
+  } catch (error) {
+    console.log("🚀 ~ sendTeamInvite ~ error:", error)
```
use our Custom Toast notifications

uparkalau replied:
What is a rocket icon?

mr-loop-1 replied:
This is for debugging, as I mentioned the error messages are not done yet

erenfn commented on backend/src/controllers/team.controller.js:
```diff
@@ -0,0 +1,56 @@
+
+const updateTeamDetails = async (req, res) => {
+    const userId = req.user.id;
+    const { name } = req.body;
```
add type check

erenfn commented on backend/src/controllers/team.controller.js:
```diff
@@ -0,0 +1,56 @@
+
+const removeMember = async (req, res) => {
+  const userId = req.user.id;
+  const { memberId } = req.body;
```
add type check

erenfn commented on backend/src/controllers/user.controller.js:
```diff
@@ -44,15 +32,49 @@ const getUsersList = async (req, res) => {
+    if(!user) {
+      return res.status(400).json({ error: "User not found" });
+    }
+    await user.update({
```
Can you move this part to service?

erenfn reviewed (changes requested):
Can you share the requests (if you used Postman, the Postman collection) you used to test the endpoints?

erenfn commented:
> > @gorkem-bwl , so it is one app or organisation per deployment (of server and db). I will update the code soon
> 
> Sorry for the late answer. This app is supposed to be deployed on a server, which is supposed to run only for one organization.

As Gorkem mentioned, the app is currently designed to be used by a single organization, so the Team table doesn't serve a meaningful purpose at the moment. I'm not sure if we plan to extend the scope to support multiple teams in the future. If we do, this table could become useful, but it's unclear whether that will happen. @gorkem-bwl @uparkalau what do you think?

gorkem-bwl commented:
The team table is supposed to show a list of users in the team. Yes, there is only one team, however it can also have multiple users. That is the reason the team table is there.

erenfn commented:
> The team table is supposed to show a list of users in the team. Yes, there is only one team, however it can also have multiple users. That is the reason the team table is there.

We have 'Users' table which shows the list of the users. It is not team specific as we only have one team. Right now the team table only stores team name

> const Team = sequelize.define(
      "Team",
      {
        id: {
          type: DataTypes.INTEGER,
          primaryKey: true,
          autoIncrement: true,
        },
        name: {
          type: DataTypes.STRING(50),
          allowNull: false,
        },
        createdAt: {
          type: DataTypes.DATE,
          allowNull: false,
          defaultValue: DataTypes.NOW,
        },
      },

gorkem-bwl commented:
Right, I think we are good to go in this scenario!

mr-loop-1 commented:
@erenfn 
To store the single value of team name in db, I thought about two options, `teams` table with a single row or `configs` key-value table with a teamName key. 
I choose first for simplicity, can also go with other one or something you could suggest.

erenfn commented:
> @erenfn To store the single value of team name in db, I thought about two options, `teams` table with a single row or `configs` key-value table with a teamName key. I choose first for simplicity, can also go with other one or something you could suggest.

Okay, I think it is fine for now. Can you share your Postman collection (if any) so that I can test the endpoints?

mr-loop-1 commented:
@erenfn I was away, will share soon.

uparkalau commented on backend/src/service/user.service.js:
```diff
@@ -0,0 +1,83 @@
+const settings = require("../../config/settings");
```
Review all constant, leave only that you are using. Double check if file name has a space, like user .service.js

uparkalau reviewed (dismissed):
Please review coderabbit refactoring suggestions and implement that really makes sense.

mr-loop-1 commented:
@erenfn 
https://api.postman.com/collections/38807808-03f17c08-2558-4273-a996-c0d3454b768f?access_key=PMAT-01J9E4W1E1CWM3CY6NAEZ2QZTR

mr-loop-1 commented:
@erenfn 
https://api.postman.com/collections/38807808-03f17c08-2558-4273-a996-c0d3454b768f?access_key=PMAT-01J9E5M1VD8VF8GN36GXAFHGXR

mr-loop-1 commented:
@erenfn @uparkalau @gorkem-bwl 
There were some suggestions regarding validating the incoming request params/body. 
This is something I was aware of while coding too but didn't do that as I think this should be done through validating middleware/dto by some lib like `express-validator`. The reason being validations such as userId, user's name etc are repeated across controllers and would be heavily duplicated.

We can add them now like done in Banner and Popup controllers if required, **or** leave as it is and create a new issue to add a validating system to the app.

mr-loop-1 commented:
> @erenfn @uparkalau @gorkem-bwl There were some suggestions regarding validating the incoming request params/body. This is something I was aware of while coding too but didn't do that as I think this should be done through validating middleware/dto by some lib like `express-validator`. The reason being validations such as userId, user's name etc are repeated across controllers and would be heavily duplicated.
> 
> We can add them now like done in Banner and Popup controllers if required, **or** leave as it is and create a new issue to add a validating system to the app.

Alongside, we can now address the error responses. I delayed because I was unsure if use the `src/utils/error.js` file or do directly. There is an argument called `errorCode` which is passed as below.

https://github.com/bluewave-labs/bluewave-onboarding/blob/e302238b01aadfdafe6c505c7ae79fe226619ecc/backend/src/utils/errors.js#L22
https://github.com/bluewave-labs/bluewave-onboarding/blob/e302238b01aadfdafe6c505c7ae79fe226619ecc/backend/src/controllers/popup.controller.js#L64-L66
I wanted to know if this is how it is supposed to be done. I can then map error codes with error messages.

gorkem-bwl commented:
Sure, we can discuss. Do you have a specific question here for me?

mr-loop-1 commented:
@gorkem-bwl I wanted to know what is meant by signup page getting invisible. Because even if a user is invited, it is through mail, and we still need name and password and hence require signup. and where does invite code fits in all this.

gorkem-bwl commented:
> @gorkem-bwl I wanted to know what is meant by signup page getting invisible. Because even if a user is invited, it is through mail, and we still need name and password and hence require signup. and where does invite code fits in all this.

Sure, let me explain. The initial signup page and the signup page that is shown after an invite are totally different in terms of what they serve.

- **Initial signup page:** This is only for admins, and will be shown only once to the user who deploys the system. It won't be visible afterwards and will never be used again. 
- **Invitation signup page:** When an invitation is initiated by the admin, the user gets and email, and this page is shown. This is an invite only signup page, and will be shown to everyone who gets an invite by the admin. 

Is it more clear now? Let me know if you need more info and/or anything is not clear.

mr-loop-1 commented:
@erenfn @uparkalau 
any inputs/approach

erenfn commented:
> @erenfn @uparkalau @gorkem-bwl There were some suggestions regarding validating the incoming request params/body. This is something I was aware of while coding too but didn't do that as I think this should be done through validating middleware/dto by some lib like `express-validator`. The reason being validations such as userId, user's name etc are repeated across controllers and would be heavily duplicated.
> 
> We can add them now like done in Banner and Popup controllers if required, **or** leave as it is and create a new issue to add a validating system to the app.

Sure I'll create a new issue for it

erenfn commented on backend/src/controllers/team.controller.js:
```diff
@@ -0,0 +1,56 @@
+
+const changeRole = async (req, res) => {
+  const userId = req.user.id;
+  const { memberId, role } = req.body;
```
Oh, okay. But the middleware still needs to check that the role of the executor is admin. Currently, I don't see any check for that.

erenfn replied:
const user = await User.findOne({
                where: {id: userId}
            });
            if(!settings.team.removeUser.includes(user.role)) {
                throw new Error("User not authorized to delete this user");
            }

This part creates repetition. It's better to create a separate middleware within authMiddleware. This will make the code cleaner and prevent unnecessary calls to the controller if the user is not an admin.

erenfn replied:
> Each route may have different set of valid roles, hence we can provide this array as a callback to the common middleware. IMO, we should handle this in the validation pr since that is also related to middleware.

That middleware will be separate from the authMiddleware. I see no issue with creating the admin role within the authMiddleware in this PR, as this is the PR where we are implementing the admin role. As I mentioned, it is a better practice to check authorization in the middleware rather than in the service. Therefore, it would be better if you could refactor that in this PR.

erenfn replied:
> Can you provide some insights on Gorkem's comment above regarding Initial signup page etc. I think the code would be highly opinionated and best if we can discuss beforehand how to proceed. We can even have a separate pr as that is more on frontend side.

I think he is saying that once this app is installed on the server, it will only allow a single signup for the admin (using our current signup page). After the admin signs up, we should disable this signup page.

What I also understood is that we will create a new signup page for members who receive an invite.

However, this issue is not related to this PR, so we can skip it for now.

gorkem-bwl replied:
> > Can you provide some insights on Gorkem's comment above regarding Initial signup page etc. I think the code would be highly opinionated and best if we can discuss beforehand how to proceed. We can even have a separate pr as that is more on frontend side.
> 
> I think he is saying that once this app is installed on the server, it will only allow a single signup for the admin (using our current signup page). After the admin signs up, we should disable this signup page.
> 
> What I also understood is that we will create a new signup page for members who receive an invite.
> 
> However, this issue is not related to this PR, so we can skip it for now.

Exactly this.

uparkalau commented on backend/src/routes/team.routes.js:
```diff
@@ -0,0 +1,23 @@
+
+router.get("/details", authenticateJWT, getTeamDetails);
+
+router.post("/invite", authenticateJWT, sendTeamInvite);
```
if we do update use PUT. For new records use POST

mr-loop-1 replied:
a separate pr was created for this recently

mr-loop-1 commented:
Its up for review now @erenfn @uparkalau

erenfn commented on backend/src/middleware/auth.middleware.js:
```diff
@@ -19,8 +20,8 @@ const authenticateJWT = async (req, res, next) => {
     }
-
-    req.user = decoded;
+    const user = await User.findOne({ where: {id: decoded.id } })
```
Why did you have to change this?

mr-loop-1 replied:
accessGuard middleware needs role of user to check if user is authorized for the action. We can't put role in jwt as it may have changed in between.

erenfn replied:
What I was trying to say is that as far as I can remember `decoded` is already the user. So doing `const user = await User.findOne({ where: {id: decoded.id } })` may be redundant. We may have to double check it

mr-loop-1 replied:
The decoded contains the jwt payload that we supplied during login/signup. It is id and email. For role I mentioned it could be outdated so can't take from there.

erenfn commented on backend/src/middleware/accessGuard.middleware.js:
```diff
@@ -0,0 +1,10 @@
+const accessGuard = (permissions) => {
+  return (req, res, next) => {
+    if(!permissions.includes(req.user.role)) {
+        return res.status(403).json({ error: "User not have required access level"});
+    }
```
Why did you mark this as resolved?  I think it is a reasonable suggestion

mr-loop-1 replied:
I already did the changes locally, didnt push yet as the bot will again flood will comments.

erenfn commented on backend/src/controllers/team.controller.js:
```diff
@@ -0,0 +1,73 @@
+
+const teamService = new TeamService();
+
+const getTeamDetails = async (req, res) => {
+  try {
+    const data = await teamService.getTeam();
+    const result = {
+        name: data.team.name,
+        users: data.users.map((user)=> ({
+            name: user.name,
+            email: user.email,
+            role: settings.user.roleName[user.role]
+        })),
+    }
+    return res.status(200).json(result);
+  } catch (err) {
+    const { statusCode, payload } = internalServerError(
+      "GET_TEAM_ERROR",
+      err.message,
+    );
+    res.status(statusCode).json(payload);
+  }
+};
```
this too

mr-loop-1 replied:
did it

erenfn replied:
it is still not pushed

mr-loop-1 replied:
I did it. The link here is not updated.

erenfn commented on backend/src/service/team.service.js:
```diff
@@ -0,0 +1,95 @@
+                throw new Error("User can't remove itself through team list");
+            }
+            
+            await User.destroy({
```
It is better to first check if the user exists before deleting them. Handling the case where the user doesn't exist in the catch block isn't the optimal approach

mr-loop-1 replied:
the auth middleware first fetchs the user so controller is called only if it exists

erenfn replied:
But isn't this a different user? The auth middleware only checks if the user is the one deleting it, not the user being deleted.

mr-loop-1 replied:
Oh yes, overlooked that.

erenfn commented on backend/src/controllers/team.controller.js:
```diff
@@ -0,0 +1,73 @@
+  const userId = req.user.id;
+  const { memberId, role } = req.body;
+  try {
+    await teamService.updateUserRole(userId, memberId, role);
```
Are you checking whether the input role is indeed a valid role and not something else?

mr-loop-1 replied:
that would be for the validator Pr to implement.

erenfn commented on backend/src/controllers/user.controller.js:
```diff
@@ -1,58 +1,88 @@
   }
 };
 
+const updateUserDetails = async (req, res) => {
+  const userId = req.user.id;
+  const inputs = req.body;
+  try {
+    await userService.updateUser(userId, inputs);
+
+    return res.status(200).json({ message: "User updated successfully" });
+  } catch (err) {
+    const { statusCode, payload } = internalServerError(
+      "UPDATE_USER_ERROR",
+      err.message,
+    );
+    res.status(statusCode).json(payload);
+  }
+}
```
this too

mr-loop-1 replied:
I think update shouldn't be the one to return the new user object as frontend's auth provider will only call current-user api.

mr-loop-1 replied:
The function in that will refresh the user object in frontend will always call normal current-user api. So, returning new user is unnecessary here. We only need the success message for Toast notification,

erenfn commented on backend/src/controllers/auth.controller.js:
```diff
@@ -2,26 +2,54 @@
     if (existingUser) return res.status(400).json({ error: "Email already exists" });
 
+    const userCount = await User.count();
     const hashedPassword = await bcrypt.hash(password, 10);
-    const newUser = await User.create({ name, surname, email, password: hashedPassword });
+
+    let newUser;
+    if(userCount) {
+      const transaction = await sequelize.transaction();
+      const invite = await Invite.findOne({
+        where: { invitedEmail: email }
+      })
+      if(!invite) {
+        throw new Error("No Invite Found");
+      }
+      
+      try {
+        await invite.destroy({ transaction });
+        newUser = await User.create({ name, surname, email, password: hashedPassword, role: invite.role }, { transaction });
+        await transaction.commit();
+      }
+      catch(err) {
+        transaction.rollback();
+        throw new Error("Error registering user by invite");
+      }
+    }
+    else {
+      newUser = await User.create({ name, surname, email, password: hashedPassword, role: settings.user.role.admin });
+    }
```
this might be also useful

mr-loop-1 replied:
did the 1st suggestion. transaction is not needed for else block as it is single independent call

erenfn commented on backend/src/routes/team.routes.js:
```diff
@@ -0,0 +1,26 @@
+
+router.get("/details", authenticateJWT, getTeamDetails);
+
+router.post("/invite", authenticateJWT, accessGuard(teamPermissions.invite), sendTeamInvite);
+router.put("/update", authenticateJWT, accessGuard(teamPermissions.update), updateTeamDetails);
+router.put("/change-role", authenticateJWT, accessGuard(teamPermissions.changeRole), changeRole);
+
+router.delete("/remove", authenticateJWT, accessGuard(teamPermissions.removeUser), removeMember);
```
this too

mr-loop-1 replied:
member id would be in body as anybody can attempt to call the route on anybody.

erenfn replied:
Okay, makes sense

mr-loop-1 commented on backend/src/controllers/team.controller.js:
```diff
@@ -0,0 +1,73 @@
+  }
+};
+
+const updateTeamDetails = async (req, res) => {
+    const userId = req.user.id;
+    const { name } = req.body;
+    try {
+      await teamService.updateTeam(userId, name);
+      return res.status(200).json({ message: "Team Details Updated Successfully" });
+    } catch (err) {
+      const { statusCode, payload } = internalServerError(
+        "UPDATE_TEAM_ERROR",
+        err.message,
+      );
+      res.status(statusCode).json(payload);
+    }
+};
```
2 and 3 are done. 1st tbd in validation pr.

mr-loop-1 commented:
@erenfn I did the changes.

erenfn commented on backend/src/middleware/accessGuard.middleware.js:
```diff
@@ -1,7 +1,7 @@
 const accessGuard = (permissions) => {
   return (req, res, next) => {
     if(!permissions.includes(req.user.role)) {
```
make this
 if(!req.user || !permissions.includes(req.user.role)) {

erenfn commented on backend/src/controllers/auth.controller.js:
```diff
@@ -2,19 +2,42 @@ const bcrypt = require("bcryptjs");
+      const invite = await Invite.findOne({
+        where: { invitedEmail: email }
+      })
+      if(!invite) {
+        throw new Error("No Invite Found");
+      }
```
This suggestion was also crucial. We shouldn't throw any errors; it crashed the server.

mr-loop-1 replied:
I will change it. But it shouldn't have crashed the server.

erenfn commented on backend/src/controllers/auth.controller.js:
```diff
@@ -2,26 +2,54 @@ const bcrypt = require("bcryptjs");
+      }
+      catch(err) {
+        transaction.rollback();
+        throw new Error("Error registering user by invite");
```
same here don't throw an error

erenfn reviewed (commented):
i get 

```
{
    "error": "User does not have required access level"
}

```
Although I am the first user

```
"user": {
        "id": 1,
        "name": "new_user23",
        "surname": "name",
        "email": "new_user23@example.com"
    },
```

erenfn commented on backend/src/controllers/auth.controller.js:
```diff
@@ -2,26 +2,54 @@ const bcrypt = require("bcryptjs");
-    const newUser = await User.create({ name, surname, email, password: hashedPassword });
+
+    let newUser;
+    if(userCount) {
```
i get 

```
{
    "error": "User does not have required access level"
}

```
Although I am the first user

```
"user": {
        "id": 1,
        "name": "new_user23",
        "surname": "name",
        "email": "new_user23@example.com"
    },
```

Check the login here

mr-loop-1 replied:
on which api?

erenfn commented on backend/src/service/user.service.js:
```diff
@@ -0,0 +1,103 @@
+      }
+    }
+
+    async deleteUser(userId) {
```
Also I get this responese when I try to delete myself as the admin with 1 userCount:

`{
    "error": "Internal Server Error",
    "message": "Error deleting user ~ relation \"invites\" does not exist",
    "errorCode": "DELETE_USER_ERROR"
}`

mr-loop-1 replied:
Did you do the migration

erenfn replied:
I did this using Sequelize sync. I think the reason we are seeing this issue is not due to migration but because we don't have an initial invite for the first user.

I believe this is related to the previous issue. I registered as the first user, but my role is 'member' and I don't have an invite. I think this should be corrected by registering this user as an admin.

In fact, right now, I think we should register all users as admins. We'll change that later.

gorkem-bwl replied:
> I did this using Sequelize sync. I think the reason we are seeing this issue is not due to migration but because we don't have an initial invite for the first user.
> 
> I believe this is related to the previous issue. I registered as the first user, but my role is 'member' and I don't have an invite. I think this should be corrected by registering this user as an admin.
> 
> In fact, right now, I think we should register all users as admins. We'll change that later.

This is fine.

erenfn commented:
After making these changes, please test your endpoints again. Testing is really really really important

mr-loop-1 commented:
> After making these changes, please test your endpoints again. Testing is really really really important

I did test all of them before pushing

erenfn commented:
> > i get
> > ```
> > {
> >     "error": "User does not have required access level"
> > }
> > ```
> 
> btw on which api call are you getting this error. I can check if this is a bug that got through

At every endpoint that requests 'admin' role.

mr-loop-1 commented:
> I did this using Sequelize sync. I think the reason we are seeing this issue is not due to migration but because we don't have an initial invite for the first user.

but why do we need this sync. i read it is not recommended for production.

erenfn commented:
> > I did this using Sequelize sync. I think the reason we are seeing this issue is not due to migration but because we don't have an initial invite for the first user.
> 
> ok, but why do we need this sync.

we don't need it is just an easy way to migrate

erenfn commented:
I think the problem is in middleware:
When I register I can see that my role is now admin. But I still get `"error": "User does not have required access level"` whenever I try to call an admin endpoint

erenfn commented:
Also when I run the code with sequelize sync I get this:

`bluewave-onboarding-db-1                  | 2024-10-13 23:41:00.109 UTC [36] ERROR:  syntax error at or near "1" at character 106
bluewave-onboarding-db-1                  | 2024-10-13 23:41:00.109 UTC [36] STATEMENT:  DROP TYPE IF EXISTS "public"."enum_users_role"; DO 'BEGIN CREATE TYPE "public"."enum_users_role" AS ENUM(1, 2); EXCEPTION WHEN duplicate_object THEN null; END';`

which indicates that there is an issue with the role enum in the constants.js

erenfn commented:
When I run with migrations, all of the endpoints work. 

I guess there is a discrepancy between the code and the migrations files

erenfn reviewed (approved):
Now it works
